### ISSUE TITLE
Images: Attribution & Deletion

### ISSUE THREAD

**Issue by Megachile** [jeffdc/gallformers#43]
When adding an image to be associated with a gall, there should be a way to also add its origin. Two possible ways to handle this, maybe not mutually exclusive:

For images from a text source, add an image uploader to the Map Species & Source page. 

Create an independent image uploader admin page that allows users to associate a new image with an existing source or add either an attribution by text (eg if the image was obtained with permission from an individual) and/or link to an iNaturalist, Bugguide, etc observation where it was obtained from (with the benefit of providing additional ecological info).

**Comment by jeffdc** [comment]
I have this stuff in the database schema but currently there is no UI for it. Adding the ability to add images from multiple places is not that hard. I think it makes sense to do it both ways you describe. I can also easily disable the ability to add an image from the main species pages. It would be a PITA to have to try and add attribution UI there.

**Comment by Megachile** [comment]
Yeah I figured that adding them on the main species page would not be convenient to add this additional info. If we can't edit other info from there anyway, I see no reason we need to add images there.

**Comment by jeffdc** [comment]
Adding to the work for this (as it is related):

- [ ] Add a way to delete images
- [ ] Add a way to re-associate an image to a different species (not required but might be convenient)

**Comment by jeffdc** [comment]
I am working on this and wanted to get some thoughts/feedback.

I think that we should have the following ways to manipulate images:
- add/delete/edit metadata via new Image Admin (see below for details)
- add/delete via species-source mapping admin page (this will automatically map the image to that source)
- display on species details pages (gall and host)

The Image Admin page will look something like this:
![Image Admin Page](https://user-images.githubusercontent.com/39919/105905143-1855d480-5ff0-11eb-8e4f-cc4c71347c44.png)

From there you will be able to view all images assigned to a species, add new images, select for deletion one or more images, and edit image metadata. The edit piece will happen via a modal that looks like this:

![Image Metadata Edit Modal](https://user-images.githubusercontent.com/39919/105905358-5226db00-5ff0-11eb-96b5-959d66d3b77a.png)

The fields shown are the current fields as they exist in the database right now. The uploader field is automatically populated at time of upload and is not editable.

Currently the other fields are just strings. We could allow the source field to map to an actual Source but then that would preclude it from being anything other than that. Maybe what we want is the option to map to a Source and the the option to just type in a text source.

**Comment by Megachile** [comment]
I think the Source field has to map to actual sources because that's the only way to cause the dropdown menu to appear and autocomplete the titles, right? I think the solution is to make mapping a Source optional, so if the image is from an iNat user or elsewhere we can just put a hyperlink instead, maybe in a new field? I think we should try to encourage one or the other--it's hard to imagine a case where we'd be putting up a photo that doesn't occur either in a primary source or somewhere else online, where there will be more info about the observation. 

If we have a link field, we could use it for images from Source mappings too, in addition to the page link for the text description (this would be convenient in that they're often separated by many pages in the old sources, where the plates are all in the back). Not sure how we'd present that on the Map Species and Source page. I guess we'd just need a second link field?

**Comment by jeffdc** [comment]
Here is what I have now:

Edit Image Details
![Edit Image Details](https://user-images.githubusercontent.com/39919/106074756-6e9f4200-60da-11eb-9a56-3f2aa77a98d9.png)

Selecting a Source from those that are already mapped to the species
![Selecting a Source from those that are already mapped to the species](https://user-images.githubusercontent.com/39919/106074797-7b239a80-60da-11eb-88aa-7652bfd503fc.png)

Hover over tips to help guide the way
![Hover over tips to help guide the way](https://user-images.githubusercontent.com/39919/106074866-9c848680-60da-11eb-8968-afd0fecdc28c.png)

Thoughts?

**Comment by Megachile** [comment]
So eg you could add a picture from a source on BHL, then put the source link to the specific page below? The "and/or" is a bit ambiguous to me. It looks like you've got all the fields there though. What's the difference between Creator and Attribution?

**Comment by jeffdc** [comment]
The thought is that you would normally only add one or the other of "Source" or "Source Link". Source being an internal reference to existing source that is defined in the system and Source Link being an external URL. I have been struggling with how to convey the and/or nature of this in the UI. 

Attribution is an open-ended field for adding attribution notes, creator would be a named individual or link to an individual.

**Comment by jeffdc** [comment]
I will also be adding the ability to add an image from the Species-Source mapping page that will (I think) not have all of this detail.

**Comment by Megachile** [comment]
My impulse would just be to clarify this with more text--maybe there's a more elegant way to do it with web design and I'm open to that but in the meantime I'd just break this into two blocks and say:

"If your image is from a publication:" and include these two fields. Only this part would be replicated on the Species-Source mapping page.
Source 
Direct Link to Image Page (if available, eg at BHL or HathiTrust. Do not duplicate main source-level link or link to a pdf):

then "If your image is not from a publication:" and include the fields you have now. I would alter the tooltip on Source link to say "Link to the original image location" though, since we're looking for the observation, not the literal hyperlink to the image itself.

**Comment by jeffdc** [comment]
Here is what I have now:

![Screen Shot 2021-01-31 at 1 37 36 PM](https://user-images.githubusercontent.com/39919/106394224-840db800-63c9-11eb-8733-a1f78167f854.png)

Behind the scenes, the "Direct Link to Image in Publication" and "Observation Link" are the same field so they can not both exist. I will add logic to the form that will prevent both from being filled out at the same time. My thoughts are, if a Source is selected then the Observation Link field will be disabled.

**Comment by Megachile** [comment]
That looks good although I think we should encourage ppl to fill out the bottom parts with an iNat or BG link. They're not really optional, even if they are technically available through the other site. That info could go down, etc, and we should credit photographers on our end too. 

Also I can't tell if you've already coded it this way but license should be a dropdown, there's only a few possibilities for that.

**Comment by jeffdc** [comment]
There are actually an infinite number of options for photo licensing since anyone can write a license and grant it for a photo that they own copyright to. What I will do is come up with a list of licenses that we can support and only allow those. I think it is a subset of CC licenses plus Public Domain.

**Comment by Megachile** [comment]
I suppose that's true but if 8 are good enough for iNaturalist then those 8 are good enough for us. Their list is:
No license (all rights reserved)
Attribution-NoDerivs
Attribution-NonCommercial NoDerivs
Attribution-NonCommercialShareAlike
Attribution-NonCommercial
Attribution
No Copyright (CC0)

**Comment by jeffdc** [comment]
I disagree. Those CC-N* licenses are troublesome. Until we get an entity other than me to take on legal responsibility for the hosting of the site I am only allowing:

- Public Domain / CC0
- CC-BY

There was a thread on the iNat forum recently that went over some of the issues with the CC-N* licenses: https://forum.inaturalist.org/t/a-case-for-changing-the-default-license-to-not-include-a-nc-clause/18690

I do not think that we can use CC-Attribution-NoDerivs since I have to scale the images so that we do not get killed with bandwidth costs on large images.

Basically, Creative Commons is a good idea, but is not well-executed and we have a very troubled and litigious society.

**Comment by Megachile** [comment]
Ok, even simpler. 

What kind of entity do we need to take legal responsibility? Isn't that why I incorporated the non-profit? Is there something else I need to do on that end? 

How does this work for licensed images (ie not CC or public domain) we individually obtain permission to use though?

**Comment by jeffdc** [comment]
For individual permission we can track it in the Attribution field for now.

Currently the Digital Ocean, AWS, and Github, accounts are all setup under me personally. We would need to setup up accounts for all of these under the org and the transfer the resources. It is a good amount of work from the tech side. I want to stop using Digital Ocean. As part of that work, #48, I will setup a new AWS account and we can make that one owned by the org. Then I can figure out how to transfer the stuff that is in my AWS account over to that one.

As for GitHub I think we would need to create a user that represents the org. Then move the project over to that user away from me. Another alternative is to create an Org in GitHub, but that costs $4/user/month which quickly adds up.

I will look into all of this stuff more once we get the site to a stable point.

**Comment by Megachile** [comment]
So just to be clear, this is what you have in mind? If I got permission from a user to use a licensed image, I would select All Rights Reserved in the License field and write "Photog Name (Used with permission)" in the Attribution field?

Let me know if you need any info about the org or if there's anything else you'd like my input on. I bet we could raise a small chunk of money per month on Patreon to cover monthly expenses if we found some that seemed really worthwhile.

**Comment by jeffdc** [comment]
Yes, that is what I am thinking.

Ok RE Org. I think what I will need from you, eventually, is that we will need to use contact info from the org etc. for setting up accounts. I will figure out when the time comes and let you know.

**Comment by Megachile** [comment]
You mentioned having host relations pre-populate in #41--is that a relatively easy thing to do? I think it will get pretty tedious adding the same rights and attribution for dozens of images from the same source (exacerbated by the fact that I want to use the consistent name formatting whenever the same person appears in multiple fields so I am looking it up each time at the moment), so it seems like having those pre-populate as well and then allowing the user to replace them where needed would be very helpful. But if it's not easy to do then it's not a priority.

**Comment by Megachile** [comment]
Similarly, you can select multiple images and delete them at once, but that doesn't seem to let you fill in the source/attribution info for multiple images at once. Is that something that can be easily changed?

**Comment by jeffdc** [comment]
The multi-select Edit should be doable without too much issue, though I have to think about how to show that you have multiple images selected when in the edit modal.

I can look at also having the fields prep-populate from data that exists for a pre-existing source-species-image.
